hil/ext/switches/ovs.py

```diff
@@ -0,0 +1,266 @@
+# governing permissions and limitations under the License.
+"""A switch driver for OpenVswitch.
+"""
+import re
+import logging
+import schema
+import os
+import ast
+import subprocess
+import shlex
+
+from hil.model import db, Switch, BigIntegerType, SwitchSession
+from hil.errors import SwitchError
+
+logger = logging.getLogger(__name__)
+
+# Class layout
+# 1. Public methods:
+# 2. Private methods
+# 3. Unimplemented superclass methods.
+
+
+class Ovs(Switch, SwitchSession):
+    """ Driver for openvswitch. """
+    api_name = 'http://schema.massopencloud.org/haas/v0/switches/ovs'
+    dir_name = os.path.dirname(__file__)
+    __mapper_args__ = {
+        'polymorphic_identity': api_name,
+    }
+
+    id = db.Column(
+            BigIntegerType, db.ForeignKey('switch.id'), primary_key=True
+            )
+    hostname = db.Column(db.String, nullable=False)
+    username = db.Column(db.String, nullable=False)
+    password = db.Column(db.String, nullable=False)
+
+    @staticmethod
+    def validate(kwargs):
+        """Checks input to match switch parameters."""
+        schema.Schema({
+            'username': basestring,
+            'hostname': basestring,
+            'password': basestring,
+        }).validate(kwargs)
+
+    @staticmethod
+    def validate_port_name(port):
+        """ This driver accepts any string as port name."""
+
+# 1. Public Methods:
+
+    def get_capabilities(self):
+        """Provides the features of this switch that HIL supports. """
+        return['Virtual_switch_for_development_Purpose_only']
+
+    def ovs_connect(self, command_string):
+        """Interacts with the Openvswitch.
+
+        Args:
+            command_string (str) : shell commands required to make changes to
+                                   openvswitch
+        Raises: SwitchError
+        Returns: If successful returns None else logs error message
+        """
+        args = shlex.split(command_string)
+        try:
+            subprocess.check_call(args)
+        except subprocess.CalledProcessError as e:
+            logger.error('%s', e)
+            raise SwitchError
+
+    def revert_port(self, port):
+        """Resets the port to the factory default.
+        Args:
+            port: Valid switch port
+
+        Returns: if successful returns None else error message.
+        """
+        import pdb; pdb.set_trace()
+
+        shell_cmd_1 = 'sudo ovs-vsctl del-port {port}'.format(port=port)
+        shell_cmd_2 = 'sudo ovs-vsctl add-port {switch} {port}'.format(
+                    switch=self.hostname, port=port
+                    ) + ' vlan_mode=native-untagged'
+        args_1 = shlex.split(shell_cmd_1)
+        args_2 = shlex.split(shell_cmd_2)
+        try:
+            subprocess.check_call(args_1)
+            subprocess.check_call(args_2)
+        except subprocess.CalledProcessError as e:
+            logger.error('%s', e)
+            raise SwitchError
```

Comment:
call the `ovs_connect` method to execute the command.

Reply:
next commit fixes it.